File: feature-map/src/main/java/home/oleg/placesnearme/feature_map/di/PlacesMapFragmentComponent.java
package home.oleg.placesnearme.feature_map.di;

import dagger.BindsInstance;
import dagger.Component;
import home.oleg.placesnearme.AppApiProvider;
import home.oleg.placesnearme.api.AppApi;
import home.oleg.placesnearme.feature_map.ui.VenuesMapFragment;
import home.oleg.placesnearme.scopes.FeatureScope;

@FeatureScope
@Component(dependencies = AppApi.class, modules = {
        VenuesMapFragmentModule.class,
        ResourceModule.class})
public interface PlacesMapFragmentComponent {

    void inject(VenuesMapFragment target);

    @Component.Builder
    interface Builder {
        PlacesMapFragmentComponent.Builder appComponent(AppApi appApi);

        @BindsInstance
        PlacesMapFragmentComponent.Builder bind(VenuesMapFragment mapFragment);

        PlacesMapFragmentComponent build();
    }

    final class Injector {
        private Injector() {
        }

        public static void inject(VenuesMapFragment fragment) {
            AppApi appApi = AppApiProvider.Initializer.INSTANCE.getAppApi(fragment);
            DaggerPlacesMapFragmentComponent.builder()
                    .appComponent(appApi)
                    .bind(fragment)
                    .build()
                    .inject(fragment);
        }
    }

}

File: feature-map/src/main/java/home/oleg/placesnearme/feature_map/di/PlacesMapFragmentComponent.kt
package home.oleg.placesnearme.feature_map.di

import dagger.BindsInstance
import dagger.Component
import home.oleg.placesnearme.coredata.di.RepoApi
import home.oleg.placesnearme.coredi.module.ViewModelFactoryModule
import home.oleg.placesnearme.coredi.scopes.FeatureScope
import home.oleg.placesnearme.corettools.di.ToolsApi
import home.oleg.placesnearme.feature_map.presentation.ui.VenuesMapFragment

@FeatureScope
@Component(
        dependencies = [ToolsApi::class, RepoApi::class],
        modules = [ViewModelFactoryModule::class, VenuesMapFragmentModule::class, ResourceModule::class])
interface PlacesMapFragmentComponent {

    fun inject(target: VenuesMapFragment)

    @Component.Builder
    interface Builder {
        fun toolsComponent(toolsProvider: ToolsApi): PlacesMapFragmentComponent.Builder

        fun repoComponent(repoApi: RepoApi): PlacesMapFragmentComponent.Builder

        @BindsInstance
        fun bind(mapFragment: VenuesMapFragment): PlacesMapFragmentComponent.Builder

        fun build(): PlacesMapFragmentComponent
    }

    object Injector {

        fun inject(fragment: VenuesMapFragment) {
            val toolsApi = ToolsApi.getInstance(fragment.activity!!.applicationContext)
            val repoApi = RepoApi.getInstance(fragment.activity!!.applicationContext)

            DaggerPlacesMapFragmentComponent.builder()
                    .toolsComponent(toolsApi)
                    .repoComponent(repoApi)
                    .bind(fragment)
                    .build()
                    .inject(fragment)
        }
    }

}
